clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinator.java

```diff
@@ -770,7 +772,8 @@ public void handle(OffsetFetchResponse response, RequestFuture<Map<TopicPartitio
                     log.debug("Group {} failed to fetch offset for partition {}: {}", groupId, tp, error.message());
 
                     if (error == Errors.UNKNOWN_TOPIC_OR_PARTITION) {
-                        future.raise(new KafkaException("Partition " + tp + " may not exist or user may not have Describe access to topic"));
+                        future.raise(new KafkaException("Partition " + tp + " may not exist or the user may not have " +
```

Comment:
@ewencp Thanks for the comments. Apologies for the long reply, but I've been thinking a bit about this as well after reviewing KIP-88. One interesting thing I found is that we actually only check topic existence when fetching offsets from zookeeper (offset fetch version 0). When fetching from Kafka, we return INVALID_OFFSET with no error code set if the user has Describe access to a topic which does not exist. So currently, the only way to get UNKNOWN_TOPIC_OR_PARTITION is if the user doesn't have Describe access to the topic. That probably explains why the current behavior has not caused any issues (that I'm aware of). 

Another related issue: when a topic is deleted, we do not remove corresponding offsets from the caches. Given the behavior above, that means the offsets remain fetchable. Generally the impact of this is minor (the offsets will eventually expire), but it could be bad if the topic was re-created quickly. In the normal case, if a consumer assigned itself a non-existing partition which had a still fetchable offset, then we'd get stuck in a loop waiting for the consumer to fetch the topic metadata. If we fixed the behavior above, then we should probably let the consumer continue trying to fetch offsets as long as it receives UNKNOWN_TOPIC_OR_PARTITION errors. Another (small) problem if we don't also fix the deletion problem is that the group cannot be cleaned up until all offsets have expired, yet the user would have no way to tell which offsets remain because the offset fetch would also return the unknown topic error. (By the way, if I recall correctly, there was a patch to fix the deletion problem, but the contributor wasn't too responsive. Probably worth trying to unblock that work.)

As to the general problem of the client not being able to distinguish between non-existing topics and authorization failures.... I agree that it sucks, but I don't see much we can do about it given the way the clients currently handles unknown topics. My hope is that we can deprecate the "retry forever until the topic exists" behavior over time as auto-creation of topics itself is deprecated. In the short term, maybe bumping up the logging is all we can do. I'm actually guessing it's uncommon in practice to not have Describe access to topics, but who knows?

Anyway, I think there are enough related issues, that we should probably deal with this in a separate JIRA, so I'm inclined to merge this patch.